Add host_id to netw::msg_addr

bhalevy opened the pull request:
This series adds a host_id member to netw::msg_addr so it can be used to identify nodes by their host_id in addition to their ip address.

The CLIENT_ID rpc verb is extended here to optionally include the host_id.
And that is used in this series to pass the node operations coordinator as a `netw::msg_addr` rather than `gms::inet_address`.

To facilitate that, the local host_id is stored in fb_utilities.

Refs #6403

denesb commented on message/messaging_service.cc:
```diff
@@ -149,16 +149,22 @@ const size_t PER_SHARD_CONNECTION_COUNT = 2;
 bool operator==(const netw::msg_addr& x, const netw::msg_addr& y) noexcept {
-    // Ignore cpu id for now since we do not really support shard to shard connections
-    return x.addr == y.addr;
+     // Ignore cpu id for now since we do not really support shard to shard connections
```
Accidental change?

bhalevy replied:
yup, will clean up

denesb commented on message/messaging_service.cc:
```diff
@@ -149,16 +149,22 @@ const size_t PER_SHARD_CONNECTION_COUNT = 2;
 }
 
-bool operator<(const netw::msg_addr& x, const netw::msg_addr& y) noexcept {
+std::strong_ordering msg_addr::operator<=>(const msg_addr& o) const noexcept {
```
Is `operator==` (above) still necessary if there is a spaceship operator available?

bhalevy replied:
yes, see https://en.cppreference.com/w/cpp/language/default_comparisons
```
The three-way comparison function (whether defaulted or not) is called whenever values are compared using <, >, <=, >=, or <=> and overload resolution selects this overload.

The equality comparison function (whether defaulted or not) is called whenever values are compared using == or != and overload resolution selects this overload.
```

denesb replied:
ok

denesb commented on message/msg_addr.hh:
```diff
@@ -10,18 +10,22 @@
-    friend bool operator<(const msg_addr& x, const msg_addr& y) noexcept;
-    friend std::ostream& operator<<(std::ostream& os, const msg_addr& x);
+    locator::host_id host_id;
+    msg_addr(gms::inet_address ip, uint32_t cpu = 0, std::optional<locator::host_id> opt_id = std::nullopt) noexcept
```
Why not ditch `std::optional` and default to `null_id` instead?

bhalevy replied:
That's possible.

IIRC initially I planned to pass an optional that might be disengaged from `messaging_service::get_source`
but `retrieve_auxiliary_opt` returns a pointer, not optional so this turned out to be a moot point.

denesb commented on locator/topology.hh:
```diff
@@ -87,6 +88,10 @@ public:
     static std::string to_string(state);
 
+    netw::msg_addr msg_addr() const noexcept {
+        return netw::msg_addr(_endpoint, 0, _host_id);
```
s/`0`/`current_shard()`/
Or do we completely ignore the shard part?

bhalevy replied:
While initializing to this_shard_id() makes sense, the default value is currently 0
and we currently ignore it anyhow.

denesb replied:
ok, no point to init it to `current_shard()` in one place if it is ignored/zeroed everywhere else

denesb commented on main.cc:
```diff
@@ -1159,6 +1159,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 return sys_ks.start(snitch.local());
             }).get();
             cfg->host_id = sys_ks.local().load_local_host_id().get0();
+            utils::fb_utilities::set_host_id(cfg->host_id);
```
I really wish we didn't pile onto this `fb_utilities` hack, but we can fix it another time.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/657/

bhalevy commented:
In v2 ([4cd75e8](https://github.com/scylladb/scylladb/commit/4cd75e8c593ecc270aacb3601f94a87acbb7735c)):
- netw: msg_addr: provide fmt::formatter
  - keep also `operator<<` and move to `namespace std` (Cc @tchaikov)
  - this was required for boost test printing for e.g. BOOST_TEST_EQUAL
- dropped `topology: node: add msg_addr getter`
  - will submit with first use
- netw: msg_addr: add host_id
  - added unit test
  - made `operator==` and `operator<=>` distinguish between msg addressed with same ip address, one with and the other without host_id
    - otherwise they collide as keys in sets/maps
    - callers need to check the presence of host_id to determine what to do if one has it and the other doesn't.
  - dropped `std::optional<host_id>` from ctor, set default value to null host_id instead
  - added convenience ctor with `(inet_address, host_id)`, as the cpu_id is unused
- message: client_info: send and retrieve host_id via CLIENT_ID message
  - select msg_addr ctor based on the ptr that `cinfo.retrieve_auxiliary_opt<locator::host_id>("host_id")` returns;

denesb commented:
> * keep also `operator<<` and move to `namespace std` (Cc @tchaikov)

Indeed, this will prove problematic in our quest to ditch `operator<<`. @tchaikov do you have a plan for this?

avikivity commented:
> This series adds a host_id member to netw::msg_addr so it can be used to identify nodes by their host_id in addition to their ip address.
> 
> The CLIENT_ID rpc verb is extended here to optionally include the host_id. And that is used in this series to pass the node operations coordinator as a `netw::msg_addr` rather than `gms::inet_address`.
> 
> To facilitate that, the local host_id is stored in fb_utilities.
> 
> Refs #6403

Why not replace it? Are not all call sites ready?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/678/

bhalevy commented:
> Since both are provided, what do we do with the host id? just ignore it?

This series just prepares the stage for followups that will make use of the host_id in node ops, then the gossiper interface.  We need both address and host_id for transitioning incrementally.

tchaikov commented:
> > * keep also `operator<<` and move to `namespace std` (Cc @tchaikov)
> 
> Indeed, this will prove problematic in our quest to ditch `operator<<`. @tchaikov do you have a plan for this?

i think to ditch `operator<<` is a side effect of the plan. the general idea is to use fmtlib for formatting, and to use its generic facilities as much as possible instead of relying on our homebrew solutions. since we don't want to two parallel systems for formatting, i think the direction is to replace `operator<<` with `fmt::format` (or `std::format` ).  to encourage developers to implement `fmt::formatter`, or, in other words, to prevent them from adding more `operator<<`, i think the best way forward is to remove them. morever, removing `operator<<` in the bottom-up approach helps us to identify the dependencies on `operator<<`, and to remove them systematically. as you know, that's what i've been working on recently.

if there is still needs of `operator<<` , in this case, from boost::test, i think, ideally, we should limit its scope. so only tests can access them. and an, IMHO, better, alternative is https://www.boost.org/doc/libs/1_82_0/libs/test/doc/html/boost_test/test_output/test_tools_support_for_logging/testing_tool_output_disable.html#ref_log_output_custom_customization_point . but as @bhalevy pointed out, developers' time is precious . probably we should not force them to identify these use cases and add the customization point for testing. so i don't think we should force our developers to be a purist. but to me, i will continue pursuing the goal by

- translating the `operator<<` to `fmt::format`
- removing the former as much as possible
- if time permits, i will use https://www.boost.org/doc/libs/1_82_0/libs/test/doc/html/boost_test/test_output/test_tools_support_for_logging/testing_tool_output_disable.html#ref_log_output_custom_customization_point to replace `operator<<` .

tchaikov commented on gms/inet_address.hh:
```diff
@@ -69,9 +69,9 @@ public:
-    }
+    friend inline auto operator<=>(const inet_address& x, const inet_address& y) noexcept {
+        return x.bytes() <=> y.bytes();
+    } 
```
trailing space.

tchaikov commented on gms/inet_address.hh:
```diff
@@ -69,9 +69,9 @@ public:
-    friend inline bool operator<(const inet_address& x, const inet_address& y) noexcept {
-        return x.bytes() < y.bytes();
-    }
+    friend inline auto operator<=>(const inet_address& x, const inet_address& y) noexcept {
```
no need to mark this function friend. as `bytes()` is a public function.

bhalevy replied:
true, I just preserved the existing declaration style

tchaikov commented on message/messaging_service.cc:
```diff
@@ -1301,8 +1295,17 @@ void messaging_service::register_node_ops_cmd(std::function<future<node_ops_cmd_
 
 } // namespace net
+
+namespace std {
```
why shall we put the operator into `std` namespace? IMHO, we just need to put it into `new` namespace. as ADL looks up the operator defined in the argument's namespace.

bhalevy replied:
you mean `netw` namespace, right?
alright with me, I wasn't aware of that.

tchaikov replied:
sorry, typo. yeah, i meant `netw`. the first parameter is in the `std` namespace, that's why ADL works also, but i think semantically, we intend to define an operator for `netw::msg_addr` not for the first parameter. hence this suggestion. see also https://en.cppreference.com/w/cpp/language/adl

bhalevy commented:
In v3 ([ca6954f](https://github.com/scylladb/scylladb/commit/ca6954fead2f2d87db3536bebfe43e257d3cee45)):
- everywhere: unalias netw::messaging_service::msg_addr
  - also in gossiper
- netw: msg_addr: provide fmt::formatter
  - keep operator<< in namespace netw (https://github.com/scylladb/scylladb/pull/13532#discussion_r1169500723)
- netw: msg_addr: use `#` delimiter to separate cpu_id from addr
  - new patch, makes printout clearer
- gms: inet_address: modernize operator== and operator<=>
  - define unary operator== and operator<=>
  - no need to friend anything (https://github.com/scylladb/scylladb/pull/13532#discussion_r1169499293)
  - remove stray trailing whitespace (https://github.com/scylladb/scylladb/pull/13532#discussion_r1169498855)

tgrabiec commented on message/msg_addr.hh:
```diff
@@ -10,18 +10,25 @@
+    msg_addr(gms::inet_address ip, uint32_t cpu = 0, locator::host_id host_id = locator::host_id::create_null_id()) noexcept
+        : addr(ip)
+        , cpu_id(cpu)
+        , host_id(host_id)
```
Can "ip" be optional? That is, can one create msg_addr with just the host_id set? Part of https://github.com/scylladb/scylladb/issues/6403 was to allow ip-less messaging.

bhalevy replied:
Until the transition is done, I don't think we can always get the ip address from the host_id in all paths, in particular in the gossiper, so I'm afraid we still need to send the ip address.
I suggest that the next step, after the transition is complete, should be to ignore the ip address and even reset it on the destination side to prove it's not needed, and only then it'd be safe to drop it on the client side.

tgrabiec replied:
We may still need to have ip-based addressing, but we should make clear what are the semantics of msg_addr in the messaging layer. For example, if I have msg_addr with both host id and the ip set, is the message guaranteed to be delivered to the given host id? This is not clarified in this patch. 

I don't think we want semantics like "try to deliver to host id, but if not found, deliver to the IP", because that kills the safety of addressing by host id, and algorithms are not guaranteed to be correct.

We could make msg_addr hold variant<inet_address, host_id>.

tgrabiec commented on message/messaging_service.cc:
```diff
@@ -380,19 +380,23 @@ messaging_service::messaging_service(config cfg, scheduling_config scfg, std::sh
         ci.attach_auxiliary("baddr", broadcast_address);
         ci.attach_auxiliary("src_cpu_id", src_cpu_id);
         ci.attach_auxiliary("max_result_size", max_result_size.value_or(query::result_memory_limiter::maximum_result_size));
+        ci.attach_auxiliary("host_id", host_id.value_or(locator::host_id::create_null_id()));
```
Why do we need host_id on the server side?

The idea in https://github.com/scylladb/scylladb/issues/6403 was to have it on the client side.

gleb-cloudius replied:
To know where the message is coming from.

tgrabiec replied:
That makes sense.

But what I was thinking here was that if we want to address by host_id, we need to know the host id of the server on the client side. This is the part which this series does not address. Basically, the host id in msg_addr is ignored on the sending side, right? In that case, what's the value of adding it?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/691/

gleb-cloudius commented on api/messaging_service.cc:
```diff
@@ -46,7 +45,7 @@ future_json_function get_client_getter(sharded<netw::messaging_service>& ms, std
         auto get_shard_map = [f](messaging_service& ms) {
             std::unordered_map<gms::inet_address, unsigned long> map;
-            ms.foreach_client([&map, f] (const msg_addr& id, const shard_info& info) {
+            ms.foreach_client([&map, f] (const netw::msg_addr& id, const shard_info& info) {
```
Add comment here on the commit message because github is stupid.

It looks like the next patch makes `std::operator<<` use `fmt::formatter` and not the other way around like the commit message implies. Also I do not understand why exactly `fmt::formatter` has the problem with the alias while the operator hadn't, but may be less intrusive way to fix it is to add another formatter for the alias?

bhalevy replied:
I think that the namespace problems were caused by me moving operator<< to namespace std instead of keeping it in the netw namespace, but I also encountered a similar issue with the finding std::hash when e.g. class gossiper had a `using msg_addr = netw::addr`.
I'll go over the commit message regarding the formatter to clarify it

bhalevy replied:
not sure.  I need to check if moving the hash to the netw namespace solves this problem.

bhalevy replied:
> But now you are moving it back to netw namespace, no? So may be the whole patch can be dropped?

dropped in v5

bhalevy commented:
In v4 (f436e102736abf369a5455005d7434bdd7ee47db):
- netw: msg_addr: provide fmt::formatter
  - clarified git commit message
- messaging_service: use relaxed msg_addr equal_to function for clients_map

bhalevy commented:
In v5 ([a319613](https://github.com/scylladb/scylladb/commit/a31961332fb8de6ca6faacaaa6175dea477765fa)):
- dropped "everywhere: unalias netw::messaging_service::msg_addr"
- introduced msg_addr::legacy_equal_to for easier backward compatibility of sets/maps during the conversion process to use host_id

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/710/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/710/

```
[2023-04-18T19:35:15.347Z] hudson.remoting.ProxyException: groovy.lang.MissingPropertyException: No such property: results for class: WorkflowScript
```

Hopefully will be solved by https://github.com/scylladb/scylla-pkg/pull/3375

bhalevy commented:
v6 ([45e7daa](https://github.com/scylladb/scylladb/pull/13532/commits/45e7daa219a154cb99976596e5340d9fe0db0a64)): 
- netw: msg_addr: make ctor(inet_address) explicit

avikivity commented:
> > Since both are provided, what do we do with the host id? just ignore it?
> 
> This series just prepares the stage for followups that will make use of the host_id in node ops, then the gossiper interface. We need both address and host_id for transitioning incrementally.

What if they conflict? And which one do you pick?

I understand an incremental transition if it's a variant<host_id, ip_address>, but when both fields are present, it's just confusing.

mykaul commented:
@bhalevy - the operator team is kinda waiting for this functionality, for their multi DC operator work. I assume there's more work around host ID conversion in core?

bhalevy commented:
> @bhalevy - the operator team is kinda waiting for this functionality, for their multi DC operator work. I assume there's more work around host ID conversion in core?

There is still work to do around identifying nodes by their host_id.
How is is going to be used by the operator team (and is there anything specific in this issue?)

mykaul commented:
> > @bhalevy - the operator team is kinda waiting for this functionality, for their multi DC operator work. I assume there's more work around host ID conversion in core?
> 
> There is still work to do around identifying nodes by their host_id. How is is going to be used by the operator team (and is there anything specific in this issue?)

@tnozicka can explain.

tnozicka commented:
I don't think we have bumped into the issues yet too much since we are working on switching the operator to use the different networking options that give us ephemeral IPs, which will trigger the issues, if any.

One case I recall we have bumped so far was having to lookup HostIds->IPs to find out token ranges
https://github.com/scylladb/scylla-operator/blob/a1ee047b53d0bfe8758224ed5fc3e0132ee078ba/pkg/controller/sidecar/sync.go#L123-L148

@zimnx can probably give you a more detailed report in general since he has been playing with it.

bhalevy commented:
Not needed anymore, closing
